File: Stack & Queue/leetcode_42.py
# Problem Link ====>>https://leetcode.com/problems/trapping-rain-water/description/

def trap_water(height):
    n = len(height)
    stack = []  # Stack to store indices of elevation bars
    water = 0   # Total trapped water

    for i in range(n):
        while stack and height[i] > height[stack[-1]]:
            bottom = stack.pop()  # Index of the bottom bar

            if not stack:
                break  # No left boundary, cannot trap water

            left = stack[-1]  # Index of the left boundary
            width = i - left - 1  # Width of the trapped water
            bounded_height = min(height[left], height[i]) - height[bottom]
            water += width * bounded_height

        stack.append(i)  # Push the current index onto the stack

    return water


# Example usage
elevation_map = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
trapped_water = trap_water(elevation_map)
print(f"Trapped water: {trapped_water} units")

File: Stack & Queue/leetcode_42.cpp

// Problem Link ====>>https://leetcode.com/problems/trapping-rain-water/description/

#include <bits/stdc++.h>
using namespace std;

int trapWater(vector<int> &height)
{
    int n = height.size();
    stack<int> st; // Stack to store indices of elevation bars
    int water = 0; // Total trapped water

    for (int i = 0; i < n; ++i)
    {
        while (!st.empty() && height[i] > height[st.top()])
        {
            int bottom = st.top(); // Index of the bottom bar
            st.pop();

            if (st.empty())
            {
                break; // No left boundary, cannot trap water
            }

            int left = st.top();      // Index of the left boundary
            int width = i - left - 1; // Width of the trapped water
            int boundedHeight = min(height[left], height[i]) - height[bottom];
            water += width * boundedHeight;
        }

        st.push(i); // Push the current index onto the stack
    }

    return water;
}

int main()
{
    vector<int> elevationMap = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    int trappedWater = trapWater(elevationMap);
    cout << "Trapped water: " << trappedWater << " units" << endl;
    return 0;
}
